lib/iris/experimental/ugrid/__init__.py

```diff
@@ -3274,6 +3279,133 @@ def context(self):
 PARSE_UGRID_ON_LOAD = ParseUGridOnLoad()
 
 
+def _meshes_from_cf(cf_reader):
+    """
+    Common behaviour for extracting meshes from a CFReader.
+
+    Simple now, but expected to increase in complexity as Mesh sharing develops.
+
+    """
+    # Mesh instances are shared between file phenomena.
+    # TODO: more sophisticated Mesh sharing between files.
+    # TODO: access external Mesh cache?
+    mesh_vars = cf_reader.cf_group.meshes
+    meshes = {
+        name: _build_mesh(cf_reader, var, cf_reader.filename)
+        for name, var in mesh_vars.items()
+    }
+    return meshes
+
+
+def load_mesh(uris, var_name=None):
```

Comment:
https://github.com/SciTools/iris/pull/4262#discussion_r692854595 nullifies this concern, yes?

Reply:
Sorry, not quite correct....
> if they all contain the same (shared) Mesh, load_meshes would return a dict with a single key matching the arg, but the value will be a list of duplicates of the Mesh (assuming they all had the same)

Should actually be 
"if they all contain the same (shared) Mesh, then load_meshes would return a key per file, and each value will be a duplicate of the single Mesh (assuming they all had the same)"

Reply:
> unexpected

Actually deliberate(!).

I wrote it this way specifically because we haven't [yet] implemented `Mesh` sharing. Part of the acceptance criteria was to return a list of `Mesh`es from a collection of files. But without `Mesh` sharing, we know that plenty of use cases would return a list of duplicate `Mesh`es, which would be confusing and might appear erroneous. Instead, I wanted to provide better context by mapping each `Mesh` to its source file.

I'm struggling with design philosophy here. One should write code for the world as it currently is, which suggests I did the write thing to help clarify outputs in the absence of `Mesh` sharing. But if we anticipate `Mesh` sharing in future, the correct API then will be different - probably returning a `list` instead of a `dict`, but that basically means I am consciously forcing a future breaking change.

---

`load_mesh()` is as close as reasonably possible to `load_cube()`, and despite the different cases it still has some of `load_cube()`'s utility, specifically the ability to extract a single named `Mesh` from a collection of files by using the `var_name` parameter. But perhaps for every bit of utility there is a greater bit of potential confusion?

Reply:
TL;DR I'm on the fence!

Reply:
> I wanted to provide better context by mapping each Mesh to its source file.

Absolutely no disagreement with that.  I think `load_meshes` is fine just as it is.
I just thought that `load_mesh` might require a bit of a re-think, as it isn't quite what I expected.

But actually I can see the aim, of checking that there is only one Mesh in multiple files, even though we can't actually do that yet.
As it already says in the docstring, "Create a single :class:`Mesh` object from one or more NetCDF files."
In that case, rather than restricting to a single input, we can leave the args as-is and document the intent.

So, thinking a bit more, I think we can generate a more "future-proof" way of actually doing this by just collecting all the meshes from the lists-of-meshes in dict values, and using a set to remove duplicates.
I.E. just change "result = [mesh for file in meshes_result.values() for mesh in file]",
to "result = **set(**[mesh for file in meshes_result.values() for mesh in file]**)**"

That uses on Mesh.\_\_eq__, so it will work with mesh-sharing as at present, but also across files once we fix mesh comparison.

Reply:
bb30d2ec40b7b0912b6b1312bb9491c50923a3d4